### PR TITLE
design(aws-ecs): tagging support for ecs (WIP)

### PR DESCRIPTION
**Pull request by iamhopaul123**
ECS tags propagation support and ECS managed tags support, see #3420.

Please read the [contribution guidelines](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md) and follow the pull-request checklist.

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by iamhopaul123** [comment]
@pkandasamy91 please review

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+Additionally, CDK support tagging and can cascade tags to all its tagable children (see [here](https://docs.aws.amazon.com/cdk/latest/guide/tagging.html)). The current CDK tagging API is shown below:
+
+``` ts
+myConstruct.node.applyAspect(new Tag('key', 'value'));
+
+myConstruct.node.applyAspect(new RemoveTag('key', 'value'));
+```
+
+As we can see, the current tagging API is not nice and intuitive enough for users which needs to be improved.
```
Comment (RomainMuller) [inline design/aws-ecs/aws-ecs-tagging-support.md]: This comment feels un-substantiated to me. What is not nice and intuitive about this API?

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+## Code changes
+
+Given the above, we should make the following changes on ECS:
+* Support tag propagation for ECS task and ECS managed tags
+  1. Add `propagateTags` and `enableECSManagedTags` properties to `BaseServiceOptions`.
+  2. Add `propagateTaskTagsFrom` and `enableECSManagedTags` properties to `Ec2ServiceProps` and `FargateServiceProps`.
+  3. Add an enum type `PropagateTagsFromType` to support `propagateTaskTagsFrom`.
+* Change CDK Tagging API
+  1. Add a method `apply(scope, 'key', 'value')` to `Tag` class and `RemoveTag` class, which calls `applyAspect`.
```
Comment (RomainMuller) [inline design/aws-ecs/aws-ecs-tagging-support.md]: I don't think this is necessary in any way, and I don't like the proposed APIs any more than I like the current API. In particular, this is very thin sugar on top of just using `applyAspec`, which is additional API surface to maintain for a very minimal value.

File: design/aws-ecs/aws-ecs-tagging-support.md (hunk 1 above)
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Does this look good to you?

File: design/aws-ecs/aws-ecs-tagging-support.md (hunk 2 above)
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Fixed!

**Comment by iamhopaul123** [comment]
@RomainMuller Fixed tagging API changing proposal. Please review.

**Comment by pkandasamy91** [comment]
@Alisaurus @efekarakus @SoManyHs @PettitWesley

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+Additionally, CDK support tagging and can cascade tags to all its taggable children (see [here](https://docs.aws.amazon.com/cdk/latest/guide/tagging.html)). The current CDK tagging API is shown below:
+
+``` ts
+myConstruct.node.applyAspect(new Tag('key', 'value'));
+
+myConstruct.node.applyAspect(new RemoveTag('key', 'value'));
+```
+
+As we can see, the current tagging API is not nice and grammatically verbose for using, since there is no reason to expose `node` to users and `applyAspect` does not indicate anything towards tags, which leaves room for improvement. Also, users need to create two objects to add tag and remove tag which causes confusion to some degree.
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Done!

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+// Create Service
+const service = new ecs.Ec2Service(stack, "Service", {
+  cluster,
+  taskDefinition,
+  propagateTaskTagsFrom: ecs.PropagatedTagSource.TASK_DEFINITION,
+  enableECSManagedTags: true,
+});
+```
+
+# Part2: Change CDK Tagging API
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Done!

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+\*`propagateTaskTagsFrom` takes enum type `PropagatedTagSource`.\
+\*Note that the reason why we define `propagateTaskTagsFrom` is because we want to have a different name for `propagateTags` to eliminate the naming confusion.
+
+In addition, for `aws-ecs-pattern` we don’t want to bother users to make additional choice, since aws-ecs-pattern modules  should be convenient to use. As a result, `SERVICE` will be set for `propagateTaskTagsFrom` when creating an ECS service, while `true` will be set for `enableECSManagedTags`.
+
+## Code changes
+
+Given the above, we should make the following changes on ECS:
+  1. Add `propagateTags` and `enableECSManagedTags` properties to `BaseServiceOptions`.
+  2. Add `propagateTaskTagsFrom` and `enableECSManagedTags` properties to `Ec2ServiceProps` and `FargateServiceProps`.
```
Comment (pkandasamy91) [inline design/aws-ecs/aws-ecs-tagging-support.md]: You don't need to add enableECSManagedTags to EC2ServiceProps and FargateServiceProps since they inherit from BaseServiceOptions.

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+  /**
+   * Propagate tags from task definition
+   */
+  TASK_DEFINITION = 'TASK_DEFINITION',
+
+  /**
+   * Do not propagate
+   */
+  NO_PROPAGATE = 'NO_PROPAGATE'
```
Comment (pkandasamy91) [inline design/aws-ecs/aws-ecs-tagging-support.md]: Don't think this enum would be useful.

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+In addition, for `aws-ecs-pattern` we don’t want to bother users to make additional choice, since aws-ecs-pattern modules  should be convenient to use. As a result, `SERVICE` will be set for `propagateTaskTagsFrom` when creating an ECS service, while `true` will be set for `enableECSManagedTags`.
+
+## Code changes
+
+Given the above, we should make the following changes on ECS:
+  1. Add `propagateTags` and `enableECSManagedTags` properties to `BaseServiceOptions`.
+  2. Add `propagateTaskTagsFrom` and `enableECSManagedTags` properties to `Ec2ServiceProps` and `FargateServiceProps`.
+  3. Add an enum type `PropagatedTagSource` to support `propagateTaskTagsFrom`.
+
+# Part 1: Support Tag Propagation for ECS Task and ECS Managed Tags
```
Comment (pkandasamy91) [inline design/aws-ecs/aws-ecs-tagging-support.md]: You might want to rephrase this to `Support ECS Managed Tags and Tag Propagation for Tasks` (Tasks can be either ECS or Fargate)

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+The `Ec2Service` and `FargateService` constructs will have two new properties:
+
+* propagateTaskTagsFrom - Propagate tags from either ECS Task Definition or Service to Task.
+* enableECSManagedTags - Enable ECS managed tags to ECS Task.
+
+An example use case to create an EC2/Fargate service:
+
+```ts
+// Create Service
+const service = new ecs.Ec2Service(stack, "Service", {
```
Comment (pkandasamy91) [inline design/aws-ecs/aws-ecs-tagging-support.md]: I know it'll be the same for FargateService, but can you add an example usage for FargateService as well?

**Comment by pkandasamy91** [comment]
The approach looks good overall! Had a couple clarifying questions.

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+* enableECSManagedTags
+
+`propagateTags` specifies whether to propagate the tags from the task definition or the service to ECS tasks and `enableECSManagedTags` specifies whether to enable Amazon ECS managed tags for the tasks within the service. Also, for `Ec2Service` and `FargateService` we will have the corresponding two new properties exposed to users:
+
+* propagateTaskTagsFrom (`SERVICE` | `TASK_DEFINITION` | `NO_PROPAGATE`)
+* enableECSManagedTags (`true` | `false`)
+
+\*SERVICE and true are default values.\
+\*`propagateTaskTagsFrom` takes enum type `PropagatedTagSource`.\
+\*Note that the reason why we define `propagateTaskTagsFrom` is because we want to have a different name for `propagateTags` to eliminate the naming confusion.
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: I think I prefer "propagateTaskTagsFrom", which is also a consensus we reached at the team design doc review.

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+## General approach
+
+The new `BaseService` class will include two more base properties:
+
+* propagateTags
+* enableECSManagedTags
+
+`propagateTags` specifies whether to propagate the tags from the task definition or the service to ECS tasks and `enableECSManagedTags` specifies whether to enable Amazon ECS managed tags for the tasks within the service. Also, for `Ec2Service` and `FargateService` we will have the corresponding two new properties exposed to users:
+
+* propagateTaskTagsFrom (`SERVICE` | `TASK_DEFINITION` | `NO_PROPAGATE`)
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: You mean an enum named "UNDEFINED"? I think NO_PROPAGATE will be more user-friendly and intuitive to CDK users if they are familiar with ECS then even better.

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
+# AWS ECS - Tagging Support
+
+Amazon ECS supports tagging for all the ECS resources (see [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-using-tags.html)). Note that ECS provides a property of ECS service named `PropagateTags` to support propagating tags from either task definitions or services to tasks (tasks cannot be tagged using CDK without setting this parameter). Currently this property is not set in CDK so that the tags are not propagated by default.
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Nice catch!

File: design/aws-ecs/aws-ecs-tagging-support.md (hunk 5 above)
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Done!

File: design/aws-ecs/aws-ecs-tagging-support.md (hunk 7 above)
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Done!

File: design/aws-ecs/aws-ecs-tagging-support.md (hunk 8 above)
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Done.

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+`propagateTags` specifies whether to propagate the tags from the task definition or the service to ECS tasks and `enableECSManagedTags` specifies whether to enable Amazon ECS managed tags for the tasks within the service. Also, for `Ec2Service` and `FargateService` we will have the corresponding two new properties exposed to users:
+
+* propagateTaskTagsFrom (`SERVICE` | `TASK_DEFINITION` | `NO_PROPAGATE`)
+* enableECSManagedTags (`true` | `false`)
+
+\*SERVICE and true are default values.\
+\*`propagateTaskTagsFrom` takes enum type `PropagatedTagSource`.\
+\*Note that the reason why we define `propagateTaskTagsFrom` is because we want to have a different name for `propagateTags` to eliminate the naming confusion.
+
+In addition, for `aws-ecs-pattern` we don’t want to bother users to make additional choice, since aws-ecs-pattern modules  should be convenient to use. As a result, `SERVICE` will be set for `propagateTaskTagsFrom` when creating an ECS service, while `true` will be set for `enableECSManagedTags`.
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: done

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 13]
```diff
@@ -0,0 +1,122 @@
+# AWS ECS - Tagging Support
+
+Amazon ECS supports tagging for all the ECS resources (see [here](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-using-tags.html)). Note that ECS provides a property of ECS service named `PropagateTags` to support propagating tags from either task definitions or services to tasks (tasks cannot be tagged using CDK without setting this parameter). Currently this property is not set in CDK so that the tags are not propagated by default.
+
+In addition, `EnableECSManagedTags` is provided as another property so that ECS managed tags can be supported and used to tag ECS tasks. However, currently users cannot set `EnableECSManagedTags` and the default value is `false`.
+
+## General approach
+
+The new `BaseService` class will include two more base properties:
```
[the remark this answers is not in the document]
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: done!

File: design/aws-ecs/aws-ecs-tagging-support.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (pkandasamy91) [reply design/aws-ecs/aws-ecs-tagging-support.md]: Can you also update it here with NONE instead of NO_PROPAGATE

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+* propagateTags
+* enableECSManagedTags
+
+`propagateTags` specifies whether to propagate the tags from the task definition or the service to ECS tasks and `enableECSManagedTags` specifies whether to enable Amazon ECS managed tags for the tasks within the service. Also, for `Ec2Service` and `FargateService` we will have the corresponding two new properties exposed to users:
+
+* propagateTaskTagsFrom (`SERVICE` | `TASK_DEFINITION` | `NONE`)
+* enableECSManagedTags (`true` | `false`)
+
+\*`SERVICE` and `true` are default values.\
```
Comment (pkandasamy91) [inline design/aws-ecs/aws-ecs-tagging-support.md]: nitpick: Can you be explicit here with which property has which default value (i.e. propagateTaskTagsFrom has a default value of `SERVICE` and enableECSManagedTags has a default value of `true`)

File: design/aws-ecs/aws-ecs-tagging-support.md [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+  ...
+
+   /**
+   * Specifies whether to propagate the tags from the task definition or the service to the tasks in the service
+   *
+   * Valid values are: PropagateTagFromType.SERVICE or PropagateTagFromType.TASK_DEFINITION
+   *
+   * @default none
+   */
+  readonly propagateTags?: PropagatedTagSource;
```
Comment (pkandasamy91) [inline design/aws-ecs/aws-ecs-tagging-support.md]: Since you've added NONE to the enum, you should use that here.

File: design/aws-ecs/aws-ecs-tagging-support.md (hunk 14 above)
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: done!

File: design/aws-ecs/aws-ecs-tagging-support.md (hunk 15 above)
Comment (iamhopaul123) [reply design/aws-ecs/aws-ecs-tagging-support.md]: ### done!

**Comment by iamhopaul123** [comment]
> I am wondering if we can provide a better developer experience. For example, if I apply a tag to my TaskDefinition and/or to my Service in the CDK, I kind of expect this to just work. Maybe the default behavior can be “smarter”?

Problem is we are not allowed to propagate tags from ECS service and task definition simultaneously. And just as Romain mentioned, users might not want to propagate any tag to tasks, which makes it hard to tell what users really want to propagate other than leaving the choice to them.

**Comment by pkandasamy91** [comment]
Closing this PR. Please add the design doc to the feature PR.
